File: src/fors.c
#include "fors.h"
#include "address.h"
#include "utils.h"

void fors_skgen(uint8_t *out, const uint8_t sk_seed[ENN],
                const uint8_t pk_seed[ENN], uint32_t adrs[ADRS_LEN],
                const uint64_t idx) {
  uint32_t skadrs[ADRS_LEN];
  memcpy(skadrs, adrs, ADRS_TYPE_IDX * sizeof(*skadrs));
  skadrs[ADRS_TYPE_IDX] = FORS_PRF;
  skadrs[ADRS_KEYPAIR_IDX] = adrs[ADRS_KEYPAIR_IDX];
  skadrs[ADRS_TREE_HEIGHT_IDX] = 0;
  skadrs[ADRS_TREE_INDEX_IDX] = htobe32(idx);

  hash_prf(out, pk_seed, sk_seed, skadrs);
}

void fors_node(uint8_t *out, const uint8_t sk_seed[ENN], const uint64_t i,
               const uint64_t z, const uint8_t pk_seed[ENN],
               uint32_t adrs[ADRS_LEN]) {
  assert(z <= A);
  assert(i < K * (1ull << (A - z)));

  uint8_t sk[ENN];
  if (z == 0) {
    fors_skgen(sk, sk_seed, pk_seed, adrs, i);
    adrs[ADRS_TREE_HEIGHT_IDX] = 0;
    adrs[ADRS_TREE_INDEX_IDX] = htobe32(i);
    hash_f(out, pk_seed, adrs, sk);
  } else {
    uint8_t buffer[2 * ENN];
    uint8_t *lnode = buffer;
    uint8_t *rnode = buffer + ENN;

    fors_node(lnode, sk_seed, 2 * i, z - 1, pk_seed, adrs);
    fors_node(rnode, sk_seed, 2 * i + 1, z - 1, pk_seed, adrs);

    adrs[ADRS_TREE_HEIGHT_IDX] = htobe32(z);
    adrs[ADRS_TREE_INDEX_IDX] = htobe32(i);

    hash_h(out, pk_seed, adrs, buffer);
  }
}

void fors_sign(uint8_t *out, const uint8_t md[FORS_MD_LEN],
               const uint8_t sk_seed[ENN], const uint8_t pk_seed[ENN],
               uint32_t adrs[ADRS_LEN]) {
  uint64_t indices[K];
  base_2b(indices, md, FORS_MD_LEN, A, K);

  uint8_t *cout = out;

  size_t i, j;
  for (i = 0; i < K; i++) {
    fors_skgen(cout, sk_seed, pk_seed, adrs, i * (1ull << A) + indices[i]);
    cout += ENN;

    for (j = 0; j < A; j++) {
      uint64_t s = (indices[i] >> j) ^ 0x01;
      fors_node(cout, sk_seed, i * (1ull << (A - j)) + s, j, pk_seed, adrs);
      cout += ENN;
    }
  }
}

void fors_pk_from_sig(uint8_t *out, const uint8_t sig_fors[FORS_SIG_LEN],
                      const uint8_t md[FORS_MD_LEN], const uint8_t pk_seed[ENN],
                      uint32_t adrs[ADRS_LEN]) {
  uint64_t indices[K];
  base_2b(indices, md, FORS_MD_LEN, A, K);

  const uint8_t *csig = sig_fors;
  uint8_t buffer[3 * ENN];
  uint8_t root[K * ENN];
  uint8_t *croot = root;

  size_t i, j;
  for (i = 0; i < K; i++) {
    adrs[ADRS_TREE_HEIGHT_IDX] = 0;
    adrs[ADRS_TREE_INDEX_IDX] = htobe32(i * (1ull << A) + indices[i]);
    hash_f(buffer + ENN, pk_seed, adrs, csig);
    csig += ENN;

    for (j = 0; j < A; j++) {
      adrs[ADRS_TREE_HEIGHT_IDX] = htobe32(j + 1);
      if (((indices[i] >> j) & 0x01) == 0) {
        adrs[ADRS_TREE_INDEX_IDX] =
            htobe32(betoh32(adrs[ADRS_TREE_INDEX_IDX]) >> 1);
        memcpy(buffer + 2 * ENN, csig, ENN);
        hash_h(buffer + ENN, pk_seed, adrs, buffer + ENN);
      } else {
        adrs[ADRS_TREE_INDEX_IDX] =
            htobe32((betoh32(adrs[ADRS_TREE_INDEX_IDX]) - 1) >> 1);
        memcpy(buffer, csig, ENN);
        hash_h(buffer + ENN, pk_seed, adrs, buffer);
      }
      csig += ENN;
    }
    memcpy(croot, buffer + ENN, ENN);
    croot += ENN;
  }

  uint32_t forspkadrs[ADRS_LEN];
  memcpy(forspkadrs, adrs, ADRS_TYPE_IDX * sizeof(*adrs));
  forspkadrs[ADRS_TYPE_IDX] = FORS_ROOTS;
  forspkadrs[ADRS_KEYPAIR_IDX] = adrs[ADRS_KEYPAIR_IDX];
  bzero(&forspkadrs[ADRS_KEYPAIR_IDX + 1], 2 * sizeof(*adrs));

  hash_t(out, K, pk_seed, forspkadrs, root);
}

File: python/fors.py
"""
From Section 8.
"""

from address import Address, AddressType
from slh_dsa import SLHDSA
from utils import base_2b, cdiv

def fors_skgen(sk_seed: bytes, pk_seed: bytes, adrs: Address, idx: int, ctx: SLHDSA) -> bytes:
    """Generate a FORS private-key value"""
    assert idx >= 0

    skadrs = adrs.copy()
    skadrs.set_type_and_clear(AddressType.FORS_PRF)
    skadrs.set_key_pair_address(adrs.get_key_pair_address())
    skadrs.set_tree_index(idx)

    return ctx.prf(pk_seed, sk_seed, bytes(skadrs.data))

def fors_node(sk_seed: bytes, i: int, z: int, pk_seed: bytes, adrs: Address, ctx: SLHDSA) -> bytes:
    """Compute the root of a Merkle subtree of FORS public values"""
    assert i >= 0
    assert z >= 0

    if z > ctx.a or i >= ctx.k * 2**(ctx.a - z):
        return b"" # NULL

    if z == 0:
        sk = fors_skgen(sk_seed, pk_seed, adrs, i, ctx)
        adrs.set_tree_height(0)
        adrs.set_tree_index(i)
        node = ctx.f(pk_seed, bytes(adrs.data), sk)
    else:
        lnode = fors_node(sk_seed, 2*i  , z-1, pk_seed, adrs, ctx)
        rnode = fors_node(sk_seed, 2*i+1, z-1, pk_seed, adrs, ctx)
        adrs.set_tree_height(z)
        adrs.set_tree_index(i)
        node = ctx.hf(pk_seed, bytes(adrs.data), lnode + rnode)

    return node

def fors_sign(md: bytes, sk_seed: bytes, pk_seed: bytes, adrs: Address, ctx: SLHDSA) -> bytes:
    """Generate a FORS signature"""
    assert len(md) == cdiv(ctx.k * ctx.a, 8)

    sig_fors = bytes() # initialize sig_fors as a zero-length byte string
    indices = base_2b(md, ctx.a, ctx.k)
    for i in range(ctx.k):
        sig_fors += fors_skgen(sk_seed, pk_seed, adrs, i * 2**ctx.a + indices[i], ctx)

        auth = bytes()
        for j in range(ctx.a):
            s = indices[i]//(2**j) ^ 0b01
            auth += fors_node(sk_seed, i*2**(ctx.a-j) + s, j, pk_seed, adrs, ctx)
        sig_fors += auth
    return sig_fors

def fors_pk_from_sig(sig_fors: bytes, md: bytes, pk_seed: bytes, adrs: Address, ctx: SLHDSA) -> bytes:
    """Compute a FORS public key from a FORS signature"""
    assert len(sig_fors) == ctx.k * (ctx.a + 1) * ctx.n
    assert len(md) == cdiv(ctx.k * ctx.a, 8)

    indices = base_2b(md, ctx.a, ctx.k)
    node =[b"" ,b""]
    root = bytes()
    for i in range(ctx.k):
        sk = sig_fors[i*(ctx.a+1)*ctx.n : (i*(ctx.a+1)+1)*ctx.n] # sk <- sig_fors.getSK(i)
        adrs.set_tree_height(0)
        adrs.set_tree_index(i * 2**ctx.a + indices[i])
        node[0] = ctx.f(pk_seed, bytes(adrs.data), sk)

        auth = sig_fors[(i*(ctx.a+1)+1)*ctx.n : (i+1)*(ctx.a+ctx.n)*ctx.n] # auth <- sig_fors.getAUTH(i)
        for j in range(ctx.a):
            adrs.set_tree_height(j+1)
            if (indices[i]//(2**j)) % 2 == 0:
                adrs.set_tree_index(adrs.get_tree_index()//2)
                node[1] = ctx.hf(pk_seed, bytes(adrs.data), node[0] + auth[j*ctx.n:(j+1)*ctx.n])
            else:
                adrs.set_tree_index((adrs.get_tree_index()-1)//2)
                node[1] = ctx.hf(pk_seed, bytes(adrs.data), auth[j*ctx.n:(j+1)*ctx.n] + node[0])
            node[0] = node[1]
        root += node[0]
    forspkadrs = adrs.copy()
    forspkadrs.set_type_and_clear(AddressType.FORS_ROOTS)
    forspkadrs.set_key_pair_address(adrs.get_key_pair_address())
    pk = ctx.t(ctx.k, pk_seed, bytes(forspkadrs.data), root)
    return pk
